docs/design/features/greenthreads.md

```diff
@@ -0,0 +1,233 @@
+
+```
+public int Receive(byte[] buffer, int offset, int size, SocketFlags socketFlags)
+```
+method gets this bit of boilerplate added to the beginning of its implementation:
+```csharp
+if (Thread.IsGreenThread)
+{
+    ValueTask<int> vt = ReceiveAsync(buffer.AsMemory(offset, size), socketFlags);
+    return vt.IsCompleted ?
+        vt.GetAwaiter().GetResult() :
+        vt.AsTask().GetAwaiter().GetResult();
+}
+```
+In other words, the synchronous Receive checks whether it’s on a green thread, and if it is, instead delegates to the corresponding ReceiveAsync method.  That method returns a ValueTask<int>.  If that operation completes synchronously, or at least has completed by the time we check, then this “synchronous” operation just returns the result of the completed operation (or throws an exception if it failed).  If, however, the operation didn’t complete yet, the implementation blocks in the equivalent of Task.Wait().  The implementation of Task.Wait has been updated with the necessary logic to suspend this green thread until the Task completes, hooking up a continuation to the Task that will resume the green thread when the Task completes, unblocking the wait.
+This approach makes it fairly straightforward for us to instrument any synchronous API with an asynchronous counterpart, having the former delegate to the latter, albeit with a variety of implications:
+-	There is now overhead involved in the synchronous operations that we didn’t exert before and that doesn’t exist in the corresponding asynchronous method. This includes needing to check the current thread to determine whether it is a green thread, and also to pay for a Task to be created and a synchronization primitive (e.g. manual reset event) to be created as part of the wait.  Some of this is inherent to the operation, and some of it could likely be optimized in the future should it be necessary.  There are other costs that will be larger and more challenging.  For example, the previous Receive example accepted a byte[] into which to store the received data. What about the overload that takes a Span<byte> instead?  Delegating to the asynchronous implementation requires something that can be stored on the heap.  The prototype currently implements this by renting an array pool array, receiving into that, and then copying back into the span, but that has measurable overhead, in particular for synchronous operations that complete synchronously due to data already being available.  There are other possibilities to explore, such as pinning and utilizing a pooled heap-based object that can be passed around, or pinning and using some other non-Memory<byte>-based asynchronous implementation implemented to work directly in terms of a pointer/length.  Employing such options, however, will increase the complexity and scope of the work required to roll this out broadly.
```

Comment:
```suggestion
-	There is now overhead involved in the synchronous operations that we didn’t exert before and that doesn’t exist in the corresponding asynchronous method. This includes needing to check the current thread to determine whether it is a green thread, and also to pay for a `Task` to be created and a synchronization primitive (e.g. manual reset event) to be created as part of the wait.  Some of this is inherent to the operation, and some of it could likely be optimized in the future should it be necessary.  There are other costs that will be larger and more challenging.  For example, the previous Receive example accepted a `byte[]` into which to store the received data. What about the overload that takes a `Span<byte>` instead?  Delegating to the asynchronous implementation requires something that can be stored on the heap.  The prototype currently implements this by renting an array pool array, receiving into that, and then copying back into the span, but that has measurable overhead, in particular for synchronous operations that complete synchronously due to data already being available.  There are other possibilities to explore, such as pinning and utilizing a pooled heap-based object that can be passed around, or pinning and using some other non-`Memory<byte>`-based asynchronous implementation implemented to work directly in terms of a pointer/length.  Employing such options, however, will increase the complexity and scope of the work required to roll this out broadly.
```